internal/provider/incident_user_data_source.go

```diff
@@ -0,0 +1,207 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/incident-io/terraform-provider-incident/internal/apischema"
+	"github.com/incident-io/terraform-provider-incident/internal/client"
+	"github.com/samber/lo"
+)
+
+var (
+	_ datasource.DataSource              = &IncidentUserDataSource{}
+	_ datasource.DataSourceWithConfigure = &IncidentUserDataSource{}
+)
+
+func NewIncidentUserDataSource() datasource.DataSource {
+	return &IncidentUserDataSource{}
+}
+
+type IncidentUserDataSource struct {
+	client *client.ClientWithResponses
+}
+
+type IncidentUserDataSourceModel struct {
+	BaseRole    *RBACRole    `tfsdk:"base_role" json:"base_role"`
+	CustomRoles []*RBACRole  `tfsdk:"custom_roles" json:"custom_roles"`
+	Email       types.String `tfsdk:"email" json:"email"`
+	ID          types.String `tfsdk:"id" json:"id"`
+	Name        types.String `tfsdk:"name" json:"name"`
+	Role        types.String `tfsdk:"role" json:"role"`
+	SlackUserId types.String `tfsdk:"slack_user_id" json:"slack_user_id"`
+}
+
+type IncidentUserRequest struct {
+	ID types.String `tfsdk:"id"`
+}
+
+type RBACRole struct {
+	Description types.String `tfsdk:"description"`
+	Id          types.String `tfsdk:"id"`
+	Name        types.String `tfsdk:"name"`
+	Slug        types.String `tfsdk:"slug"`
+}
+
+func (i *IncidentUserDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*client.ClientWithResponses)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source User",
+			fmt.Sprintf("Expected *client.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	i.client = client
+}
+
+func (i *IncidentUserDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_user"
+}
+
+func (i *IncidentUserDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data IncidentUserDataSourceModel
+	req.Config.Get(ctx, &data)
+
+	var user *client.UserWithRolesV2
+	if !data.ID.IsNull() {
+		resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		result, err := i.client.UsersV2ShowWithResponse(ctx, data.ID.ValueString())
+		if err == nil && result.StatusCode() >= 400 {
+			err = fmt.Errorf(string(result.Body))
+		}
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read user, got error: %s", err))
+			return
+		}
+		user = &result.JSON200.User
+	} else if !data.Email.IsNull() {
+		result, err := i.client.UsersV2ListWithResponse(ctx, &client.UsersV2ListParams{
```

Comment:
I quite dislike this vs. having a lookup endpoint where you can provide an email, but I think we've had this for ages so I'll code

Reply:
I just added it today lol

Reply:
hmm. I don't love this in our public API vs. an explicit lookup endpoint. One for #api-design? Not a hill I'll die on